pkg/controller.v1alpha3/suggestion/suggestion_controller.go

```diff
@@ -0,0 +1,176 @@
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	return &ReconcileSuggestion{
+		Client:           mgr.GetClient(),
+		SuggestionClient: fake.New(),
+		scheme:           mgr.GetScheme(),
+		Composer:         composer.New(mgr.GetScheme()),
+	}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("suggestion-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	err = c.Watch(&source.Kind{Type: &suggestionsv1alpha3.Suggestion{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	err = c.Watch(&source.Kind{Type: &appsv1.Deployment{}}, &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &suggestionsv1alpha3.Suggestion{},
+	})
+	if err != nil {
+		return err
+	}
+
+	err = c.Watch(&source.Kind{Type: &corev1.Service{}}, &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &suggestionsv1alpha3.Suggestion{},
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+var _ reconcile.Reconciler = &ReconcileSuggestion{}
+
+// ReconcileSuggestion reconciles a Suggestion object
+type ReconcileSuggestion struct {
+	client.Client
+	composer.Composer
+	suggestionclient.SuggestionClient
+
+	scheme *runtime.Scheme
+}
+
+// Reconcile reads that state of the cluster for a Suggestion object and makes changes based on the state read
+// and what is in the Suggestion.Spec
+// TODO(user): Modify this Reconcile function to implement your Controller logic.  The scaffolding writes
+// a Deployment as an example
+// Automatically generate RBAC rules to allow the Controller to read and write Deployments
+// +kubebuilder:rbac:groups=apps,resources=deployments,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=apps,resources=deployments/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=katib.kubeflow.org,resources=suggestions,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=katib.kubeflow.org,resources=suggestions/status,verbs=get;update;patch
+func (r *ReconcileSuggestion) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	logger := log.WithValues("Suggestion", request.NamespacedName)
+	// Fetch the Suggestion instance
+	instance := &suggestionsv1alpha3.Suggestion{}
+	err := r.Get(context.TODO(), request.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// For additional cleanup logic use finalizers.
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+	oldS := instance.DeepCopy()
+
+	deploy, err := r.DesiredDeployment(instance)
```

Comment:
I am confused here. Why do we need a deployment and service for a suggestion?

Reply:
I guess, suggestion controller will depend on the deployment controller and service controller, and binding the deployment and the service will expose a service port in cluster.
@gaocegege Can you correct my guess.

Reply:
@johnugeorge I think we should create the suggestion first, then we can communicate with the suggestion.

For example, if an experiment uses random, we need to create a random suggestion for the experiment.